staging/src/k8s.io/apimachinery/pkg/util/cache/lruexpirecache.go

```diff
@@ -61,29 +53,127 @@ func NewLRUExpireCacheWithClock(maxSize int, clock Clock) *LRUExpireCache {
 	value      interface{}
 	expireTime time.Time
+
+	// If not nil, a computation is in progress. You must wait and then
+	// check finalValue.
+	wait              <-chan struct{}
+	finalValue        interface{}
+	finalValueOmitted bool
 }
 
 // Add adds the value to the cache at key with the specified maximum duration.
 func (c *LRUExpireCache) Add(key interface{}, value interface{}, ttl time.Duration) {
 	c.lock.Lock()
 	defer c.lock.Unlock()
-	c.cache.Add(key, &cacheEntry{value, c.clock.Now().Add(ttl)})
+	et := c.clock.Now().Add(ttl)
+	c.cache.Add(key, &cacheEntry{
+		value:      value,
+		expireTime: et,
+	})
 }
 
-// Get returns the value at the specified key from the cache if it exists and is not
-// expired, or returns false.
+// Get returns the value at the specified key from the cache if it exists and
+// is not expired, or returns false. If the key is not in the cache because it
+// is still being computed, Get will return immediately as if the key were not
+// present. To wait, use GetOrWait.
 func (c *LRUExpireCache) Get(key interface{}) (interface{}, bool) {
+	ce, _ := c.getOrBegin(key, 0)
+	if ce == nil {
+		return nil, false
+	}
+	if ce.wait != nil {
+		return nil, false
+	}
+	return ce.value, true
+}
+
+// ComputeFunc is used by GetOrWait to perform some work. You *must* exit if
+// the abort chan is closed/has a value. On failure, return a ttl of 0 to avoid
+// being cached; this will cause all waiting on the function to observe a cache miss.
+type ComputeFunc func(abort <-chan time.Time) (value interface{}, ttl time.Duration)
+
+// GetOrWait returns the value at the specified key from the cache if it exists
+// and is not expired. Otherwise, it calls the compute() function to produce
+// the value; requests for the same key that arrive within computationTime will
+// be blocked while the computation is in progress; for all such requests,
+// compute() will only be called once.
+//
+// In the unlikely event the cache has so much churn that a freshly computed
+// value can't be regotten, this will attempt to compute it again, but not more
+// than 3 times, and not if more than `computationTime` passes overall.
+func (c *LRUExpireCache) GetOrWait(key interface{}, compute ComputeFunc, computationTime time.Duration) (interface{}, bool) {
+	thisTTL := computationTime
+	if thisTTL < 0 {
+		thisTTL = 0
+	}
+	ce, ch := c.getOrBegin(key, thisTTL)
+	if ch != nil {
+		// We are the first caller for this key, so begin the
+		// computation.
+		abort := c.clock.After(thisTTL)
+		go func() {
+			defer close(ch)
+			value, ttl := compute(abort)
+			if ttl > 0 {
+				// write a new entry before closing the channel
+				// so that new getters of this key can check
+				// ce.value and not wait; existing waiters
+				// won't look at finalValue until after we
+				// close the channel, so there is no race.
+				c.Add(key, value, ttl)
+				ce.finalValue = value
+			} else {
+				c.Remove(key)
+				ce.finalValueOmitted = true
```

Comment:
If finalValue stores nil to represent omission, do we still need finalValueOmitted ?

Reply:
Someone might deliberately store a nil.